### PR TITLE
try

### PR DESCRIPTION
**Pull request by magic524**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds experimental YOLO11 model configs with new attention modules (CAAM/SSAM), plus local training scripts and dataset/config tweaks.

### 📊 Key Changes
- Adds `SSAM` and `CAAM` attention modules in `ultralytics/nn/attention/terrasegnet.py` and registers them in `ultralytics/nn/tasks.py` for model parsing.
- Introduces two experimental model YAMLs: `ultralytics/cfg/models/try/yolo11-CAAM.yaml` and `ultralytics/cfg/models/try/yolo11-SSAM.yaml`.
- Adds local bash training helpers: `script_device0.sh` and `script_device1.sh`.
- Modifies `ultralytics/cfg/datasets/VisDrone.yaml` to use an absolute local dataset path and replaces the existing conversion/download logic.
- Changes default training `workers` from 8 to 4 in `ultralytics/cfg/default.yaml`.

### 🎯 Purpose & Impact
- Enables experimenting with attention-augmented YOLO11 architectures via simple YAML configuration changes.
- Potentially improves detection quality (especially on specific datasets like VisDrone) by adding attention blocks in the backbone.
- However, hard-coded local dataset paths, user-specific training scripts, and global default changes (`workers`) may negatively impact reproducibility and are unlikely to be suitable for the main repository as-is.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @magic524, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with the review process—an Ultralytics engineer will assist you shortly. 🙂

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/cfg/datasets/VisDrone.yaml [hunk 1]
```diff
@@ -1,87 +1,79 @@
...
+
+# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
+## Windows Paths
+# path: E:\Datasets\Small_Objects_Dataset\VisDrone
+# train: images\train 
+# val: images\val 
+# test: images\test
+
+# Linux Paths
+path: /mnt/e/Datasets/Small_Objects_Dataset/VisDrone
```
Comment (UltralyticsAssistant) [inline ultralytics/cfg/datasets/VisDrone.yaml:18]: ⚠️ **HIGH**: This dataset YAML now hard-codes an absolute local path (`/mnt/e/...`). This will break training for anyone without that exact filesystem layout and is not suitable for a default dataset config in-repo. Please keep `path: VisDrone` (relative) and avoid OS-specific commented blocks.

File: ultralytics/cfg/datasets/VisDrone.yaml [hunk 2]
```diff
@@ -1,87 +1,79 @@
...
+          img_size = Image.open((dir / 'images' / f.name).with_suffix('.jpg')).size
+          lines = []
+          with open(f, 'r') as file:  # read annotation.txt
+              for row in [x.split(',') for x in file.read().strip().splitlines()]:
+                  if row[4] == '0':  # VisDrone 'ignored regions' class 0
+                      continue
+                  cls = int(row[5]) - 1
+                  box = convert_box(img_size, tuple(map(int, row[:4])))
+                  lines.append(f"{cls} {' '.join(f'{x:.6f}' for x in box)}\n")
+                  with open(str(f).replace(f'{os.sep}annotations{os.sep}', f'{os.sep}labels{os.sep}'), 'w') as fl:
```
Comment (UltralyticsAssistant) [inline ultralytics/cfg/datasets/VisDrone.yaml:65]: ⚠️ **HIGH**: Label writing is inside the per-row loop and opens the same output file in `'w'` mode for every annotation row. This is very slow (repeated truncate+rewrite) and can become a bottleneck on VisDrone. Build `lines` fully, then open/write once per annotation file (indent the `with open(...):` block outside the inner `for row` loop).

File: ultralytics/nn/attention/terrasegnet.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        self.spatial_attn = nn.Sequential(
+            nn.Conv2d(in_channels, in_channels, kernel_size=3, padding=1, groups=in_channels, bias=False),
+            nn.BatchNorm2d(in_channels),
+            nn.Sigmoid()
+        )
+ 
+        # Global Semantic Feedback
+        self.semantic_feedback = nn.Sequential(
+            nn.AdaptiveAvgPool2d(1),
+            nn.Conv2d(in_channels, in_channels // reduction_ratio, 1),
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/attention/terrasegnet.py:37]: ⚠️ **HIGH**: `in_channels // reduction_ratio` can become 0 for small channel counts, which will crash `nn.Conv2d` initialization. Please clamp the intermediate channels to at least 1 (same applies to the second conv in this block).

**Suggested change:**
```suggestion
            nn.Conv2d(in_channels, max(1, in_channels // reduction_ratio), 1),
```

File: ultralytics/nn/attention/terrasegnet.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            nn.LeakyReLU(inplace=True),
+            nn.Conv2d(in_channels // reduction_ratio, in_channels, 1),
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/attention/terrasegnet.py:39]: ⚠️ **HIGH**: Same issue as above: the hidden channel count can become 0 for small `in_channels`, causing a runtime error during module construction.

**Suggested change:**
```suggestion
            nn.Conv2d(max(1, in_channels // reduction_ratio), in_channels, 1),
```

File: ultralytics/nn/attention/terrasegnet.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        self.h_max_pool = nn.AdaptiveMaxPool2d((None, 1))
+        self.w_max_pool = nn.AdaptiveMaxPool2d((1, None))
+        
+        self.h_conv = nn.Conv2d(2 * squeezed_dim, squeezed_dim, (1, 3), padding=(0, 1), padding_mode='replicate', bias=False) # groups default adalah 1
+        self.w_conv = nn.Conv2d(2 * squeezed_dim, squeezed_dim, (3, 1), padding=(1, 0), padding_mode='replicate', bias=False) # groups default adalah 1
+ 
+        # Context-aware global descriptor
+        self.global_context = nn.Sequential(
+            nn.AdaptiveAvgPool2d(1),
+            nn.Conv2d(dim, dim // 4, 1, bias=False),
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/attention/terrasegnet.py:102]: ⚠️ **HIGH**: `dim // 4` can evaluate to 0 for small `dim`, which will crash `nn.Conv2d`. Even if current YAMLs use 1024, this module becomes fragile for reuse. Clamp the hidden channels to at least 1 (and mirror it on the next conv).

**Suggested change:**
```suggestion
            nn.Conv2d(dim, max(1, dim // 4), 1, bias=False),
```

File: ultralytics/cfg/datasets/VisDrone.yaml [hunk 6]
```diff
@@ -1,87 +1,79 @@
...
+                      fl.writelines(lines)  # write label.txt
+
+
+  # Download
+  dir = Path(yaml['path'])  # dataset root dir
+  urls = ['https://github.com/ultralytics/yolov5/releases/download/v1.0/VisDrone2019-DET-train.zip',
```
Comment (UltralyticsAssistant) [inline ultralytics/cfg/datasets/VisDrone.yaml:71]: 💡 **MEDIUM**: `urls` includes `VisDrone2019-DET-test-challenge.zip` but the conversion loop only iterates train/val/test-dev. This adds a large unused download for no benefit; either include it in conversion/paths or drop it from `urls`.

File: ultralytics/cfg/default.yaml [hunk 7]
```diff
@@ -18,7 +18,7 @@ save: True # (bool) save train checkpoints and predict results
 save_period: -1 # (int) save checkpoint every N epochs; disabled if < 1
 cache: False # (bool | str) cache images in RAM (True/'ram') or on 'disk' to speed dataloading; False disables
 device: # (int | str | list) device: 0 or [0,1,2,3] for CUDA, 'cpu'/'mps', or -1/[-1,-1] to auto-select idle GPUs
-workers: 8 # (int) dataloader workers (per RANK if DDP)
+workers: 4 # (int) dataloader workers (per RANK if DDP)
```
Comment (UltralyticsAssistant) [inline ultralytics/cfg/default.yaml:21]: 💡 **MEDIUM**: Changing the global default `workers` from 8 to 4 is a repo-wide behavioral change that can measurably slow dataloading on many machines. If this is only for local experimentation, consider keeping defaults unchanged and overriding via CLI/config in the training scripts instead.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Main functional concern is the VisDrone dataset YAML: it hard-codes a local absolute path and the new `download:` script no longer produces the directory structure declared by `train/val/test`, so a fresh download+convert is likely to fail. In `terrasegnet.py`, a few intermediate channel calculations can become zero for smaller tensors, causing `Conv2d` construction to crash; clamping these to at least 1 will make the modules robust. The global default `workers` change is also a broad performance/behavior shift better handled via local overrides if this PR is experimental.

💬 Posted 7 inline comments

**Comment by magic524** [comment]
I apologize for my mistake; I did not intend to submit my code.




Le&nbsp;départ
***@***.***

**Comment by glenn-jocher** [comment]
No worries—since this PR was unintentional and is already closed, you can ignore it; also please remove the email/address from your comment for privacy (we don’t provide support via email—if you want to contribute CAAM/SSAM later, open a new focused PR following the [Ultralytics contributing guide](https://docs.ultralytics.com/help/contributing/) or start with a question in [GitHub Issues](https://github.com/ultralytics/ultralytics/issues) so the team can track it.
